xdsl/transforms/csl_stencil_bufferize.py

```diff
@@ -379,6 +381,92 @@ def match_and_rewrite(self, op: arith.Constant, rewriter: PatternRewriter, /):
         )
 
 
+@dataclass(frozen=True)
+class InjectApplyOutsIntoLinalgOuts(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.ApplyOp, rewriter: PatternRewriter, /):
+        yld = op.done_exchange.block.last_op
+        assert isinstance(yld, csl_stencil.YieldOp)
+        new_dest: list[SSAValue] = []
+        new_yield_args: list[SSAValue] = []
+        additional_args: list[SSAValue] = []
+        to_remove: list[Operation] = []
+
+        for arg, yld_arg in zip(op.dest, yld.arguments, strict=True):
+            if (
+                not isinstance(yld_arg, OpResult)
+                or not isinstance(yld_arg.op, bufferization.ToMemrefOp)
+                or not isinstance(yld_arg.op.tensor, OpResult)
+                or not isinstance(
+                    linalg_op := yld_arg.op.tensor.op,
+                    linalg.NamedOpBase | linalg.Generic,
+                )
+                or not isa(arg_t := arg.type, AnyMemRefType)
+                or not isa(yld_arg.type, AnyMemRefType)
+            ):
+                new_dest.append(arg)
+                new_yield_args.append(yld_arg)
+                continue
+            additional_args.append(arg)
+            if len(yld_arg.uses) == 1:
+                to_remove.append(yld_arg.op)
+
+            arg = op.done_exchange.block.insert_arg(
+                arg.type, len(op.done_exchange.block.args)
+            )
+            arg_to_tensor = to_tensor_op(arg, writable=True)
+            symmetric_offsets = tuple(
+                (s - d) // 2
+                for s, d in zip(
+                    arg_t.get_shape(), yld_arg.type.get_shape(), strict=True
+                )
+            )
+
+            extract_slice_op = tensor.ExtractSliceOp(
+                operands=[arg_to_tensor, [], [], []],
+                result_types=[yld_arg.op.tensor.type],
+                properties={
+                    "static_offsets": DenseArrayBase.from_list(i64, symmetric_offsets),
+                    "static_sizes": DenseArrayBase.from_list(
+                        i64, yld_arg.type.get_shape()
+                    ),
+                    "static_strides": DenseArrayBase.from_list(i64, (1,)),
+                },
+            )
+            rewriter.insert_op(
+                [arg_to_tensor, extract_slice_op], InsertPoint.before(linalg_op)
+            )
+            rewriter.replace_op(
+                linalg_op,
+                type(linalg_op).build(
+                    operands=[linalg_op.inputs, [extract_slice_op.result]],
+                    result_types=linalg_op.result_types,
+                    regions=[linalg_op.detach_region(r) for r in linalg_op.regions],
+                    properties=linalg_op.properties,
+                    attributes=linalg_op.attributes,
+                ),
+            )
+        if len(additional_args) > 0:
```

Comment:
```suggestion
        if additional_args:
```

Reply:
Using `len` to check for whether something is empty is recommended against, as it can potentially be O(n)